README.md
# timings-docker

This repo provides **docker-compose** support for the node/express based [**TIMINGS API**](https://github.com/godaddy/timings) only! **This is not the API itself** but merely a collection of scripts and configuration files to run the API in a docker based environment. For details about the API itself, please check out the repo here: [https://github.com/godaddy/timings](https://github.com/godaddy/timings).

Also, see the FAQ section in the Wiki for more help & tips: [https://github.com/Verkurkie/timings-docker/wiki/FAQ-page](https://github.com/Verkurkie/timings-docker/wiki/FAQ-page).

## IMPORTANT

**> IF YOU ARE UPDATING YOUR INSTALLATION OF THIS REPO, PLEASE READ THE [UPDATING.md](./docs/UPDATING.md) DOCUMENT!!**

## Usage

Before jumping into the installation & usage of this repo and the associated Docker container, please read the following notes.
It is really crucial that you understand Docker and `docker-compose`, especially when upgrading this repo!!

- When you start the Docker infrastructure for the first time, a number of folders is created for logs and data files:

  ```no-lang
  [host folder]
    /elasticsearch
      /data
    /kibana
      /data
    /timings
      /logs
  ```

- As you use the API, elasticsearch data is stored on the local file system of the Docker Host! The data can be found in the `./elasticsearch/data/nodes/` folder
- When you upgrade this repo, there is a possibility that the `docker-compose.yml` file is pointing at newer versions of Elasticsearch and Kibana
  - ==>> in this scenario, please read the [UPDATING.md](./docs/UPDATING.md) document!
  - **please make sure that you have migrated your elasticsearch data when upgrading to a new MAJOR version of elasticsearch/kibana!**

Last but not least, please reference the [system requirements](#system-requirements) to check if your system meets the needs.

### Step 1. Clone it

Clone this repo to a folder of your choice:

```lang-console
$ git clone git@github.com/godaddy/timings-docker.git
Cloning into 'timings-docker'...
...
$ cd timings-docker
```

### Step 2. Configure it

It is recommended that you create a custom config file. You can copy the sample config file (`./timings-docker/timings/config/sample_default.json`) and:

- save it in a location of your choice (example: `/etc/timings/timings.json`)
- edit the file according to your needs - see also here: [https://github.com/godaddy/timings/blob/master/CONFIG.MD](https://github.com/godaddy/timings/blob/master/CONFIG.MD)

### Step 3. Run it

You should now be able to run the environment by running:

```lang-console
$ CONFIGFILE=/etc/timings/timings.json docker-compose up
WARNING: The HOSTNAME variable is not set. Defaulting to a blank string.
Starting elasticsearch ... done
Starting kibana        ... done
Starting timings       ... done
Attaching to elasticsearch, kibana, timings
timings          | [2021-03-01T09:33:45.171Z][info] - timings API - LOGGING - logs files stored in: ["/src/logs"]
timings          | [2021-03-01T09:33:45.175Z][info] - timings API - LOGGING - log level: INFO
timings          | [2021-03-01T09:33:45.514Z][info] - timings API - CONFIG - Using config ["/src/.config.js"]
timings          | [2021-03-01T09:33:45.514Z][info] - timings API - STATUS - Server v1.4.5 is running at http://08ed9a9ac258:80
timings          | [2021-03-01T09:33:45.522Z][info] - Elasticsearch - UTILS - [PORTCHECK] waiting for host [elasticsearch] at port [9200] ...
elasticsearch    | [2021-03-01T09:33:48,685][INFO ][o.e.n.Node               ] [66ca5bd94595] version[7.11.1], pid[7], build[default/docker/ff17057114c2199c9c1bbecc727003a907c0db7a/2021-02-15T13:44:09.394032Z], OS[Linux/4.19.128-microsoft-standard/amd64], JVM[AdoptOpenJDK/OpenJDK 64-Bit Server VM/15.0.1/15.0.1+9]
elasticsearch    | [2021-03-01T09:33:48,690][INFO ][o.e.n.Node               ] [66ca5bd94595] JVM home [/usr/share/elasticsearch/jdk], using bundled JDK [true]
... [more elasticsearch messages]

timings          | [2021-03-01T09:34:04.367Z][info] - Elasticsearch - UTILS - [HEALTHCHECK] status [yellow] of host [elasticsearch] is OK!
timings          | [2021-03-01T09:34:04.378Z][info] - Elasticsearch - UTILS - [INFO] found elastic v7.11.1 ...
timings          | [2021-03-01T09:34:04.426Z][info] - Elasticsearch - UTILS - [UPDATE] API and ELK are up-to-date!
```

Above example is showing the main log output messages of the [timings] service that you should look for! During the first startup, you should ultimately see a line that says `[UPDATE] API and ELK are up-to-date!` or `[TEMPLATE] created/updated [cicd-perf]` in case the API was updated! This confirms that the API waited for Elasticsearch to be healthy and it updated the template for `cicd-perf*` in Elasticsearch.

```bash
$ CONFIGFILE=/etc/timings/timings.json docker-compose up
```

**NOTE:** The first time you install this or when you use the `pull / --build` argument(s), Docker will (re-)build the containers! The output will look different and the entire process will take a bit longer to complete.

You should use `CONFIGFILE=/etc/timings/timings.json docker-compose pull && docker-compose up --build` every time the timings images is updated. This ensures you're getting the latest container!

### Step 4. Check it

After the containers have started, you can test the app by browsing to the following URLs (where `hostname` is the name of you Docker host system):

|Service|Link|
|-|-|
|timings|http://hostname
|elasticsearch|http://hostname:9200|
|kibana|http://hostname:5601|

## Other considerations

Depending on your host system, you may need to make some further adjustments or you can consider a few optional settings. Keep reading below.

### Update docker-compose file [optional]

Instead of using the `CONFIGFILE` variable to start docker-compose, you can reference your config file in the `docker-compose.yml` file:

- Update `./timings-docker/docker-compose.yml` and **uncomment + edit** the _volumes_ section for the `timings` service to map your config file to the container's `/src/config/default.json` file:

```yaml
  volumes
  volumes:
    - "${CONFIGFILE}:/src/config/default.json"   <<-- edit this line
    - ./timings/logs:/timings/logs
```

**If you don't use a custom config file, the API will use default values**. Settings such as the ElasticSearch host (`ES_HOST`), Kibana host (`KB_HOST`), etc. don't need to be included because they are already defined in the docker-compose yaml file.

### Host specific settings

Before you can run the API you may have to make a few modifications to your docker host:

#### Add user to `docker` group [Linux only]

You have to add your user account to the `docker` group and logging out & back in again. If you don't do this, you have to run `docker-compose` with `sudo` which is not recommended! You can use the following command:

```lang-console
sudo usermod -aG docker ${USER}
```

#### Custom elasticsearch data directory [optional]

This is optional. By defaults, the elasticsearch container will use the `timings-docker/elasticsearch/data` directory on the **docker host** to store its data. If you want to use a different location, you can edit the `volumes` section of the `timings-docker/docker-compose.yml` file and point to the desired location:

```yaml
  elasticsearch:
    ...
    volumes:
      - ./elasticsearch/data:/usr/share/elasticsearch/data    <<-- edit this line
    ...
```

#### Set permissions for the data & logging directories [Linux only]

You need to set the necessary read/write permissions to the data and logging directories! Depending on how you are running docker-compose, the permissions may look different

In this example, we're setting full permissions on all of the directories:

```lang-console
$ sudo chmod 777 ./elasticsearch/data
$ sudo chmod 777 ./kibana/data
$ sudo chmod 777 ./timings/logs
```

## System requirements

- Linux or Windows based OS with the following pre-requisites:

  - Docker and docker-compose
    - Docker: [https://docs.docker.com/engine/installation/](https://docs.docker.com/engine/installation/)
    - Docker-compose: [https://docs.docker.com/compose/install/](https://docs.docker.com/compose/install/)
  - Windows Subsystem for Linux (Windows only)
    - To support the `wait-for-it.sh` script (until we have a PowerShell equivalent)
    - More info: [https://msdn.microsoft.com/en-us/commandline/wsl/install-win10](https://msdn.microsoft.com/en-us/commandline/wsl/install-win10) and [https://msdn.microsoft.com/commandline/wsl/install-on-server](https://msdn.microsoft.com/commandline/wsl/install-on-server)
- Min. 4GB memory for elasticsearch (8+GB is recommended)
- Storage space for elasticsearch data
  - required amount depends on test frequency. As an indicator, running the API for ~6 mo at GoDaddy with ~40,000 tests/day produced ~170Gb of data.
  - storage can also be mounted remotely (see [here](#custom-elasticSearch-data-directory-optional))


CHANGELOG.md
## Commits on Mar 13, 2018

* docs: docs: Updated documentation & cleanup
* updated README doc
* cleaned up some old files

## Commits on Mar 12, 2018

* doc: updated documentation

## Commits on Mar 10, 2018

* Updated config instructions
* Updates for latest API version
* Changed the config file handling
* New startup for timings API (waits for elasticsearch healthy state)
* Updated documentation

## ## Commits on Feb 2, 2018

* Update documentation
* addressing issue #22
* elaborate on config file location

## Commits on Jan 6, 2018

* Updates for API v.1.1.4
* No longer need the import.py script
* Now using docker images instead of docker builds
* Elasticsearch/Kibana v.5.6.2

## Commits on Dec 12, 2017

* Update README

## Commits on Dec 11, 2017

* Change to docker compose config
* Now pulling the `timings` container as a public image instead of building it locally

## Commits on Dec 5, 2017

* Updates to import script and Kibana items

## Commits on Nov 22, 2017

* Fix issue with Python import script
* Removed reference to `usessl` argument * Updated documentation

## Commits on Nov 15, 2017

* Add ES authentication and update template
* The import script now supports basic auth to elasticsearch
* arguments for the import script have changed!
* The `cicd-perf` template was causing issues with very long URLs

## Commits on Nov 4, 2017

* Updates and fixes
* Added elasticsearch data folder with necessary permissions
* Set corret file permissions to wait-for-it.sh script
* Expanded documentation with more text & images to address user feedback

## Commits on Oct 31, 2017

* Update documentation with more detailed info and instructions

## Commits on Oct 26, 2017

* Set executable permissions for `timings/wait-for-it.sh` script

## Commits on Oct 25, 2017

* Updated documentation & kibana elements

## Commits on Oct 8, 2017

* Initial commit


upgrade/configs/.config_sample.js
const config = {
  "params": {                       // Client-side parameters
    "required": [                     // REQUIRED LOG parameters (have to start with 'log.')
      "log.test_info",                  // Information about the test-step
      "log.env_tester",                 // Environment of the test machine (i.e. `local`, `grid`, `saucelabs`, `iphone 7`, `galaxy-s6`, etc.)
      "log.team",                       // The name of the test team (this is used to organize the Kibana dashboard)
      "log.browser",                    // The browser used to perform the test (i.e. `Chrome`, `FireFox`, etc. - for API tests, use something like `api-curl`)
      "log.env_target"                  // Environment of the target application (i.e. `dev`, `test`, `prod`)
    ],
    "defaults": {                     // DEFAULT parameters (will be used if they are not provided in the client's POST body)
      "baseline": {                     // These settings are used to calculate the baseline
        "days": 7,                        // Number of days to calculate the baseline for (default: 7)
        "perc": 75,                       // Percentile to calculate (default: 75)
        "padding": 1.2                    // Multiplier to calculate extra padding on top of the baseline (default: 1.2 = 20%)
      },
      "flags": {                        // These booleans determine the output and other actions to be performed
        "assertBaseline": true,           // Whether or not to compare against baseline (default: true)
        "debug": false,                   // Request extra debug info from the API (default: false)
        "esTrace": false,                 // Request elasticsearch output from API (default: false)
        "esCreate": false,                // Save results to elasticsearch (default: false)
        "passOnFailedAssert": false       // Pass the test, even when the performance is above the threshold (default: false)
      }
    }
  }
}

module.exports = config;

elasticsearch/import/import.py
"""
Imports the kibana objects into Elasticsearch
:return: True if operation is completed successfully, False otherwise.
"""
import sys
import os
import json

import argparse
import requests

# Test parameter settings
PARSER = argparse.ArgumentParser()

# Common arguments
PARSER.add_argument(
    '--apihost', action='store', default='localhost',
    help='full hostname or IP address of the timings server (default=localhost)')
PARSER.add_argument(
    '--apiport', action='store', default='80',
    help='port of the timings server (default=80)')
PARSER.add_argument(
    '--esprotocol', action='store', default='http',
    help='scheme used by the elasticsearch server (default=http)')
PARSER.add_argument(
    '--eshost', action='store', default='localhost',
    help='full hostname or IP address of the elasticsearch server (default=localhost)')
PARSER.add_argument(
    '--esport', action='store', default='9200',
    help='port of the elasticsearch server (default=9200)')
PARSER.add_argument(
    '--esuser', action='store',
    help='username for elasticsearch - if required')
PARSER.add_argument(
    '--espasswd', action='store',
    help='The password for elasticsearch - if required')
PARSER.add_argument(
    '--kbindex', action='store', default='.kibana',
    help='the kibana index (default=.kibana)')
PARSER.add_argument(
    '--kbhost', action='store', default='localhost',
    help='full hostname or IP address of the kibana server (default=localhost)')
PARSER.add_argument(
    '--kbport', action='store', default='5601',
    help='port of the kibana server (default=5601)')
PARSER.add_argument(
    '--replace', action='store',
    help='replace `TIMINGS` with this string')


OPTIONS = PARSER.parse_args()

if not len(sys.argv) > 1:
    print(">>> You did not provide any or all of the arguments - " +
          "defaults will be used!")

IMPORT_FILE = open(os.path.join(os.path.abspath(
    os.path.dirname(__file__)), 'kibana_items.json')).read()

if OPTIONS.replace:
    IMPORT_FILE = IMPORT_FILE.replace("TIMINGS", OPTIONS.replace.upper())

IMPORT_JSON = json.loads(IMPORT_FILE)

TEMPLATE_JSON = json.load(
    open(os.path.join(os.path.abspath(
        os.path.dirname(__file__)
    ), 'cicd_template.json')))

PERF_JSON = json.load(
    open(os.path.join(os.path.abspath(
        os.path.dirname(__file__)
    ), 'sample_data.json')))

BASE_IMPORT_URL = '{0}://{1}:{2}'.format(OPTIONS.esprotocol, OPTIONS.eshost, OPTIONS.esport)

STRING = ("Starting import to server [{0}] on port [{1}] to index [{2}]".format(
    OPTIONS.eshost, OPTIONS.esport, OPTIONS.kbindex))
print(STRING)
print("=" * len(STRING))


def kb_import():
    """Import dashboards, visualizations and index-patterns into Kibana"""
    try:
        for item in IMPORT_JSON:
            _id = item["_id"]
            _type = item["_type"]
            _source = item["_source"]
            _title = _source["title"]

            if _type == 'index-pattern' and _id == 'cicd-perf':
                if OPTIONS.apiport != '80':
                    api_host = OPTIONS.apihost + ':' + OPTIONS.apiport
                else:
                    api_host = OPTIONS.apihost

                _source['fieldFormatMap'] = _source['fieldFormatMap'].replace(
                    '__api__hostname', api_host)

            response = requests.post(
                BASE_IMPORT_URL + '/' + OPTIONS.kbindex +
                '/' + _type + '/' + _id,
                data=json.dumps(_source),
                headers={'content-type': 'application/json'},
                auth=(OPTIONS.esuser, OPTIONS.espasswd)
            )

            check_response(response, 'import [' + _type + ']', _title)

    except requests.exceptions.RequestException as err:
        print(err)


def kb_default_index():
    """Set cicd-perf as the default index"""
    try:
        response = requests.put(
            BASE_IMPORT_URL + '/' + OPTIONS.kbindex + '/config/5.6.2',
            data=json.dumps({'defaultIndex': 'cicd-perf'}),
            headers={'content-type': 'application/json'},
            auth=(OPTIONS.esuser, OPTIONS.espasswd)
        )

        check_response(response, 'default index', 'cicd-perf')

    except requests.exceptions.RequestException as err:
        print(err)


def es_template():
    """Adds basic template for cicd-perf index to ElasticSearch"""
    try:
        response = requests.post(
            BASE_IMPORT_URL + '/_template/cicd-perf',
            data=json.dumps(TEMPLATE_JSON),
            headers={'content-type': 'application/json'},
            auth=(OPTIONS.esuser, OPTIONS.espasswd)
        )

        check_response(response, 'add template', 'cicd-perf')

    except requests.exceptions.RequestException as err:
        print(err)


def es_sample_data(index, data_type):
    """Adds sample data to main indices (cicd-perf and cicd-perf-res)"""
    doc_type = data_type
    if data_type == 'resource_nav' or data_type == 'resource_uri':
        doc_type = 'resource'

    try:
        response = requests.post(
            BASE_IMPORT_URL + '/' + index + '/' + doc_type + '/',
            data=json.dumps(PERF_JSON[data_type]),
            headers={'content-type': 'application/json'},
            auth=(OPTIONS.esuser, OPTIONS.espasswd)
        )

        check_response(response, 'add sample data', data_type)

    except requests.exceptions.RequestException as err:
        print(err)


def check_response(response, job, item):
    """Check the request responses and prints results"""
    job_summ = " - job: {0} - item: {1}".format(job, item)
    if response.ok:
        print("PASS - {0}".format(job_summ))
    else:
        print('FAIL - {0} - StatusCode: {1} - Reason: {2}, Error: {3}'.format(
            job_summ, response.status_code, response.reason, response.content))


if __name__ == '__main__':
    kb_import()
    kb_default_index()
    es_template()
    es_sample_data('cicd-perf', 'navtiming')
    es_sample_data('cicd-perf-res', 'resource_nav')
    es_sample_data('cicd-perf-res', 'resource_uri')
